src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -154,8 +154,16 @@ public static DateTimeOffset GetLastWriteTime(string fullPath)
             return data.ftLastWriteTime.ToDateTimeOffset();
         }
 
-        private static void MoveDirectory(string sourceFullPath, string destFullPath, bool sameDirectoryDifferentCase)
+        private static void MoveDirectory(string sourceFullPath, string destFullPath, bool isCaseSensitiveRename)
         {
+            // Source and destination must have the same root.
+            ReadOnlySpan<char> sourceRoot = Path.GetPathRoot(sourceFullPath);
+            ReadOnlySpan<char> destinationRoot = Path.GetPathRoot(destFullPath);
+            if (!sourceRoot.Equals(destinationRoot, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new IOException(SR.IO_SourceDestMustHaveSameRoot);
```

Comment:
How is this scenario handled on Unix now? According to the [manual](https://man7.org/linux/man-pages/man2/rename.2.html) `rename` can fail with `EXDEV`, but we don't have any special handling for it, so `Interop.GetExceptionForIoErrno` is going to get called and I suspect that a different error message is going to be used?

Reply:
On Unix, `Path.GetPathRoot` results in `/` for both the source and destination path, so the roots are always the same. Making this check Windows only doesn't change behavior on Unix.